codegenerator/cli/templates/static/codegen/src/EventLookup.res

```diff
@@ -0,0 +1,149 @@
+open Belt
+type eventMod = module(Types.InternalEvent)
+type contract = {
+  name: string,
+  chains: array<ChainMap.Chain.t>,
+}
+type registeredEvent<'a> = {contract: contract, event: 'a, mutable isWildcard: bool}
+type eventsAtTopic0<'a> = Single(registeredEvent<'a>) | Multiple(dict<registeredEvent<'a>>)
```

Comment:
What would IDs contain?

Reply:
The keys of entities. Maybe worth calling keys

Reply:
It's useful for:
- Saving the Object.keys call
- Cheaply get the number entities

Reply:
Ok I understand. I think you mean events not entities right?

Also we can then just rather save an int for number of events. That can work. If it's just for checking single or multiple. That sounds good to me 👍🏼

Reply:
Hmmm thinking about this:

```res
switch eventsAtTopic0 {
  | {ids: [], events} => //None case
  | {ids: [singleId], events} => //Some single case
  | {ids: multiple, events} =>//Multiple case
}
```

A think you end up with the same cases with worse semantics. Not sure if it's an improvement.

Reply:
Yes, something like this. Also, since there are no delitions doing it like this might be more convenient:
```
{
  events: array<registeredEvent<'a>>,
  eventByContractName: dict<registeredEvent<'a>>
}
```
And your example will be:
```
switch eventsAtTopic0 {
  | {events: []} => //None case
  | {events: [event]} => //Some single case
  | {events} =>//Some multiple case
}
```
> I think you end up with the same cases with worse semantics. Not sure if it's an improvement.  
I agree that for the example above the semantics a little bit worse, even though I personally find it fine. There are some benefits coming along:
- For some cases you don't need to have separate logic for Single and Multiple events, since now it's possible to do eventsAtTopic0. eventByContractName->Js.Dict.get without additional checks (also works for eventsAtTopic0.events)
- No Js.Dict.keys and Js.Dict.values calls and additional object allocations for Single and Multiple variants. In our case this is not very important, since the code runs only on events registration, but still I think it's good to do our best with performance

Reply:
Nice, cool that sounds like a good 👌🏼 will do.